services/appflowy-collaborate/src/ws2/actors/workspace.rs

```diff
@@ -0,0 +1,446 @@
+        );
+        self.last_message_id = msg.last_message_id.max(msg.last_message_id);
+        for session in self.sessions_by_client_id.values() {
+          if session.collab_origin == msg.sender {
+            continue; // skip the sender
+          }
+          session.conn.do_send(WsOutput {
+            message: ServerMessage::Update {
+              object_id: msg.object_id,
+              collab_type: msg.collab_type,
+              flags: msg.update_flags,
+              last_message_id: msg.last_message_id,
+              update: msg.update.clone(),
+            },
+          });
+        }
+      },
+      Err(err) => {
+        tracing::error!(
+          "failed to read update stream message for workpsace {}: {:?}",
+          self.workspace_id,
+          err
+        );
+        ctx.stop();
+      },
+    }
+  }
+}
+
+impl StreamHandler<(ObjectId, AwarenessStreamUpdate)> for Workspace {
+  fn handle(&mut self, (object_id, msg): (ObjectId, AwarenessStreamUpdate), _: &mut Self::Context) {
+    tracing::trace!(
+      "received awareness update for {}/{}",
+      self.workspace_id,
+      object_id
+    );
+    for (session_id, sender) in self.sessions_by_client_id.iter() {
+      if sender.collab_origin == msg.sender {
+        continue; // skip the sender
+      }
+      tracing::trace!("sending awareness update to {}", session_id);
+      sender.conn.do_send(WsOutput {
+        message: ServerMessage::AwarenessUpdate {
+          object_id,
+          collab_type: CollabType::Unknown,
+          awareness: msg.data.encode_v1().into(),
+        },
+      });
+    }
+  }
+}
+
+impl Handler<Join> for Workspace {
+  type Result = ();
+
+  fn handle(&mut self, msg: Join, ctx: &mut Self::Context) -> Self::Result {
+    if msg.workspace_id == self.workspace_id {
+      let handle = SessionHandle::new(msg.collab_origin, msg.addr.clone());
+      self.sessions_by_client_id.insert(msg.session_id, handle);
+      tracing::trace!(
+        "attached session `{}` to workspace {}",
+        msg.session_id,
+        msg.workspace_id
+      );
+      let store = self.store.clone();
+      let workspace_id = self.workspace_id;
+      let session = msg.addr;
+      if let Some(last_message_id) = msg.last_message_id {
+        ctx.spawn(
+          async move {
+            if let Err(err) =
+              Self::publish_collabs_created_since(store, workspace_id, last_message_id, session)
+                .await
+            {
+              tracing::error!(
+                "failed to send missing collabs for workspace {}: {}",
+                workspace_id,
+                err
+              );
+            }
+          }
+          .into_actor(self),
+        );
+      }
+    }
+  }
+}
+
+impl Handler<Leave> for Workspace {
+  type Result = ();
+
+  fn handle(&mut self, msg: Leave, ctx: &mut Self::Context) -> Self::Result {
+    if msg.workspace_id == self.workspace_id {
+      self.sessions_by_client_id.remove(&msg.session_id);
```

Comment:
Should we remove the client after a timeout? The server might not receive a Leave message even if the client exits.

Reply:
Why? Leave is sent by heartbeat interval or when the session actor is stopping.